config/routes.rb

```diff
@@ -61,22 +61,30 @@
           get tab, to: "activities#show", defaults: {tab: tab}
         end
       end
-      resource :level_b_activity_upload, only: [:new, :show, :update]
+      namespace :level_b do
+        namespace :activities do
+          resource :upload, only: [:new, :show, :update]
+        end
+      end
     end
 
     resources :reports, only: [:new, :create, :show, :edit, :update, :index] do
       resource :state, only: [:edit, :update], controller: :reports_state
-      resource :activity_upload, only: [:new, :show, :update]
-      resource :forecast_upload, only: [:new, :show, :update]
-      resource :actual_upload, only: [:new, :show, :update]
       get "variance" => "report_variance#show"
       get "forecasts" => "report_forecasts#show"
       get "actuals" => "report_actuals#show"
       get "budgets" => "report_budgets#show"
       get "activities" => "report_activities#show"
       get "comments" => "report_comments#show"
-      namespace :uploads do
-        resource :actual_history, only: [:new, :update]
+      namespace :activities do
+        resource :upload, only: [:new, :show, :update]
+      end
+      namespace :actuals do
+        resource :upload, only: [:new, :show, :update]
+        resource :history_upload, only: [:new, :update]
+      end
+      namespace :forecasts do
+        resource :upload, only: [:new, :show, :update]
```

Comment:
You could group the upload resources. Only one line less, so it's whether it's worth grouping similar activities or sticking to more explicit entity groupings

```
%i(activities actuals forecasts).each do |report_namespace|
  namespace report_namespace do
    resource upload, only: [:new, :show, :update]
  end
end

namespace :actuals do
  resource :history_upload, only: [:new, :update]
end
```

Not sure if it's possible to encapsulate it into a method for reuse with the budgets too, since the method might not respect the resources nesting?